examples/link-viewer-cameras.py

```diff
@@ -0,0 +1,56 @@
+"""
+Link cameras between two viewers
+================================
+
+This lets you zoom in on different layers side by side instead of overlaying
+them. [1]_
+
+.. [1] https://forum.image.sc/t/zooming-in-on-the-same-region-on-two-different-layers/95659/5
+
+.. tags:: gui, layers, visualization-advanced, multicanvas
+"""
+from functools import partial
+
+from skimage import data, filters
+
+import napari
+
+# example data
+coins = data.coins()
+binary = coins > filters.threshold_otsu(coins)
+
+# make two viewers
+viewer0 = napari.Viewer()
+viewer1 = napari.Viewer()
+
+# add a layer to each
+viewer0.add_image(coins)
+viewer1.add_labels(binary)
+
+# note: check available events with:
+# viewer0.camera.events.emitters
+
+# reusable functions to set camera zoom and center
+def set_zoom(camera, event):
+    camera.zoom = event.value
+
+def set_center(camera, event):
+    camera.center = event.value
```

Comment:
```suggestion
def set_zoom(event):
    if event.source == viewer0.camera:
        viewer1.camera.zoom = event.value
    else:
        viewer0.camera.zoom = event.value

def set_center(event):
    if event.source == viewer0.camera:
        viewer1.camera.center = event.value
    else:
        viewer0.camera.center = event.value
```

If you prefer not to have object for synchronization, maybe use this approach?

Reply:
I like this, shows also that events have useful attributes. Btw, should the `==` be `is`?

Reply:
@Czaki is this closure containing the viewers not worse for keeping refs alive?

At any rate, I personally much prefer simple functions and partials. The number of connections stays the same so I see little benefit of this approach — it increases branching in the functions, and involves writing functions that are not generally reusable. (It already hurts a bit to have to use event.value instead of just the value, but when we move all our events to psygnal we can drop that.)

Reply:
And I do not like partials until we migrate to psygnal as our handling o partials is not good. And I do not want to spend my time on improving events when they are marked for removal.

Reply:
> And I do not like partials until we migrate to psygnal as our handling o partials is not good. And I do not want to spend my time on improving events when they are marked for removal.

And this clousure is keeping reference to module level variable, so no, it does not will keep it alive. The problem is when you define a function in function. 

Maybe we could use weakref here.

Reply:
> this clousure is keeping reference to module level variable

module-level variable? I don't know what you mean here.

> our handling of partials is not good. 

but I thought it's less of a problem in this case because we only keep a reference to cameras (comment [here](https://github.com/napari/napari/pull/6881#discussion_r1590760028)), which are tiny objects?

> Maybe we could use weakref here.

No. End users should not have to know about weakrefs.

As we discussed in the other thread, we *can* make our own version of partial, but I was hoping we could do this in later PRs.

Reply:
> but I thought it's less of a problem in this case because we only keep a reference to cameras (comment [here](https://github.com/napari/napari/pull/6881#discussion_r1590760028)), which are tiny objects?

Because newcomers will not see the difference between storing reference to camera and reference to viewer. 

> As we discussed in the other thread, we _can_ make our own version of partial, but I was hoping we could do this in later PRs.

Or we may unwrap partial during connection and store weak references. 

> module-level variable? I don't know what you mean here.

Script level. Variable that is defined in `globals`, not `locals`

Reply:
> Script level. Variable that is defined in globals, not locals

Ok but the point is that *it's a bad example* — just like partial isn't bad here but could be, a closure isn't bad here but could be? ie newcomers will not appreciate this subtle issue. And honestly the same thing could be said of the lambdas???

> Because newcomers will not see the difference between storing reference to camera and reference to viewer.

Ok, well, I can add a note. But:

> Or we may unwrap partial during connection and store weak references.

this is actually quite nice — and it would mean we could just leave this example as is, right? (And I think it could also be done with lambdas? Which gets me back to where I wanted to go — users not having to worry about this stuff.)

Reply:
> Ok but the point is that _it's a bad example_ — just like partial isn't bad here but could be, a closure isn't bad here but could be? ie newcomers will not appreciate this subtle issue. And honestly the same thing could be said of the lambdas???

It is not a local function (function in function), so it is not a bad pattern. 

> this is actually quite nice — and it would mean we could just leave this example as is, right? (And I think it could also be done with lambdas? Which gets me back to where I wanted to go — users not having to worry about this stuff.)

It will also crash multiple existing code, as the plugin depends on storing hard references in partials/lambdas. 
So it will be a multiple releases migration when first create code to inspect and warn, then start storing weak references in the next release.

Reply:
> plugin depends on storing hard references

which plugin? I don't think many plugins would actually be inspecting the partials?

Reply:
I mean any plugin that uses partial in for a loop for storing complex objects. When we start converting it to weakref, then reference to such an object will be lost.

Reply:
ok but I doubt (?) that many people are relying on the partials to be their sole reference to particular objects?

Reply:
> ok but I doubt (?) that many people are relying on the partials to be their sole reference to particular objects?

I'm not sure if we removed all such things from napari codebase

See for example https://github.com/napari/napari/blob/a8c822cb2180a2307ae0b1952e4e6c36405143d8/napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py#L572